File: src/constants/elements.ts
import type { Element } from "../types";

/**
 * Element advantage cycle: Fire → Earth → Wind → Water → Fire
 * Advantage: 1.5x damage, Disadvantage: 0.67x, Neutral: 1.0x
 */
const ADVANTAGE_MAP: Record<Element, Element> = {
  fire: "earth",
  earth: "wind",
  wind: "water",
  water: "fire",
};

export function getTypeMultiplier(attacker: Element, defender: Element): number {
  if (attacker === defender) return 1.0;
  if (ADVANTAGE_MAP[attacker] === defender) return 1.5;
  if (ADVANTAGE_MAP[defender] === attacker) return 0.67;
  return 1.0;
}

export const ELEMENT_COLORS: Record<Element, string> = {
  fire: "#ff6b35",
  water: "#4fc3f7",
  earth: "#8d6e63",
  wind: "#aed581",
};

export const ELEMENT_LABELS: Record<Element, string> = {
  fire: "Fire",
  water: "Water",
  earth: "Earth",
  wind: "Wind",
};

File: crates/combat-engine/src/elements.rs
use crate::types::Element;

/// Element advantage cycle: Fire -> Earth -> Wind -> Water -> Fire
/// Returns multiplier x100: 150 = super effective, 67 = resisted, 100 = neutral
pub fn get_type_multiplier(attacker: Element, defender: Element) -> u32 {
    if attacker == defender {
        return 100;
    }

    let attacker_beats = match attacker {
        Element::Fire => Element::Earth,
        Element::Earth => Element::Wind,
        Element::Wind => Element::Water,
        Element::Water => Element::Fire,
    };

    let defender_beats = match defender {
        Element::Fire => Element::Earth,
        Element::Earth => Element::Wind,
        Element::Wind => Element::Water,
        Element::Water => Element::Fire,
    };

    if attacker_beats == defender {
        150
    } else if defender_beats == attacker {
        67
    } else {
        100
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn all_16_element_pairs() {
        use Element::*;
        // Same element = 100
        assert_eq!(get_type_multiplier(Fire, Fire), 100);
        assert_eq!(get_type_multiplier(Water, Water), 100);
        assert_eq!(get_type_multiplier(Earth, Earth), 100);
        assert_eq!(get_type_multiplier(Wind, Wind), 100);

        // Advantages = 150
        assert_eq!(get_type_multiplier(Fire, Earth), 150);
        assert_eq!(get_type_multiplier(Earth, Wind), 150);
        assert_eq!(get_type_multiplier(Wind, Water), 150);
        assert_eq!(get_type_multiplier(Water, Fire), 150);

        // Disadvantages = 67
        assert_eq!(get_type_multiplier(Earth, Fire), 67);
        assert_eq!(get_type_multiplier(Wind, Earth), 67);
        assert_eq!(get_type_multiplier(Water, Wind), 67);
        assert_eq!(get_type_multiplier(Fire, Water), 67);

        // Neutral (non-adjacent in cycle) = 100
        assert_eq!(get_type_multiplier(Fire, Wind), 100);
        assert_eq!(get_type_multiplier(Wind, Fire), 100);
        assert_eq!(get_type_multiplier(Water, Earth), 100);
        assert_eq!(get_type_multiplier(Earth, Water), 100);
    }
}
